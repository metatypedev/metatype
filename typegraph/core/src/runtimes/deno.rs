// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::runtimes as wit;

#[derive(Debug)]
pub struct MaterializerDenoModule {
    pub file: String,
}

#[derive(Debug)]
pub struct MaterializerDenoImport {
    pub func_name: String,
    pub module: wit::MaterializerId,
    pub secrets: Vec<String>,
}

#[derive(Debug)]
pub struct MaterializerDenoStatic {
    pub value: serde_json::Value,
}

#[derive(Debug)]
pub enum DenoMaterializer {
    Static(MaterializerDenoStatic),
    Inline(wit::MaterializerDenoFunc),
    Predefined(wit::MaterializerDenoPredefined),
    Module(MaterializerDenoModule),
    Import(MaterializerDenoImport),
}
