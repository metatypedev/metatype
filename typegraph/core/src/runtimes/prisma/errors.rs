// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

// use crate::wit::core::Error;

// pub fn relationship_not_found(source_model: &str, field: &str) -> Error {
//     format!("relationship target not found for  {source_model}::{field}")
// }
