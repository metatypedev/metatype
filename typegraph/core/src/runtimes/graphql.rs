use crate::wit::runtimes::{self as wit};

#[derive(Debug)]
pub enum GraphqlMaterializer {
    Query(wit::MaterializerGraphqlQuery),
    Mutation(wit::MaterializerGraphqlQuery),
}
