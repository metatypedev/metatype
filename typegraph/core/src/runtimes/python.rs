// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::wit::runtimes::{self as wit};

#[derive(Debug)]
pub enum PythonMaterializer {
    Lambda(wit::MaterializerPythonLambda),
    Def(wit::MaterializerPythonDef),
    Module(wit::MaterializerPythonModule),
    Import(wit::MaterializerPythonImport),
}
