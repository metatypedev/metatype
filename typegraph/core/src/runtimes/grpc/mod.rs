// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use std::rc::Rc;

use crate::errors::Result;
use crate::global_store::Store;
use crate::wit::core::RuntimeId;
use crate::wit::runtimes::{Effect as WitEffect, GrpcRuntimeData};
use crate::{conversion::runtimes::MaterializerConverter, typegraph::TypegraphContext};

use common::typegraph::Materializer;

use serde_json::{from_value, json};

use super::Runtime;

pub mod type_generation;

pub fn get_gprc_data(runtime_id: RuntimeId) -> Rc<GrpcRuntimeData> {
    match Store::get_runtime(runtime_id).unwrap() {
        Runtime::Grpc(data) => data,
        _ => unreachable!(),
    }
}

#[derive(Debug)]
pub struct GrpcMaterializer {
    pub method: String,
}

impl MaterializerConverter for GrpcMaterializer {
    fn convert(
        &self,
        c: &mut TypegraphContext,
        runtime_id: RuntimeId,
        effect: WitEffect,
    ) -> Result<common::typegraph::Materializer> {
        let runtime = c.register_runtime(runtime_id)?;
        let data = from_value(json!({"method": self.method})).map_err(|e| e.to_string())?;
        Ok(Materializer {
            name: "grpc".into(),
            runtime,
            effect: effect.into(),
            data,
        })
    }
}
