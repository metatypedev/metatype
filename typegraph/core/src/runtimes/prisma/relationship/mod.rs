// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::global_store::with_store;
use crate::types::Type;
use crate::types::TypeFun;
use crate::wit::core::TypeId;
use crate::wit::runtimes::Error as TgError;

mod discovery;
mod registry;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Cardinality {
    Optional,
    One,
    Many,
}

fn get_rel_name(wrapper_type: TypeId) -> Result<Option<String>> {
    with_store(|s| {
        let mut type_id = wrapper_type;

        loop {
            let ty = s.get_type(type_id)?;
            match ty {
                Type::Proxy(p) => {
                    if let Some(name) = p.data.get_extra("rel_name") {
                        return Ok(Some(name.to_string()));
                    }
                }
                _ => {
                    if let Some(wrapper_type) = ty.as_wrapper_type() {
                        type_id = wrapper_type.get_wrapped_type(s).unwrap();
                        continue;
                    } else {
                        // concrete type
                        return Ok(None);
                    }
                }
            }
        }
    })
}

#[derive(Clone)]
pub struct RelationshipSource {
    pub model_type: TypeId,
    pub wrapper_type: TypeId,
    pub cardinality: Cardinality,
}

// no wrapper type; to be determined later
#[derive(Clone)]
pub struct RelationshipTarget {
    pub model_type: TypeId,
    /// field of this model pointing to the other side of the relationship
    pub field: String,
    // /// cardinality for the other side of the relationship;
    // /// telling whether the field has a type M, M?, or M[]
    // pub cardinality: Cardinality,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RelationshipModel {
    pub model_type: TypeId,
    pub wrapper_type: TypeId,
    pub cardinality: Cardinality,
    pub field: String,
}

impl RelationshipModel {
    pub fn from_source(source: RelationshipSource, field: String) -> Self {
        Self {
            model_type: source.model_type,
            wrapper_type: source.wrapper_type,
            cardinality: source.cardinality,
            field,
        }
    }

    pub fn get_rel_name(&self) -> Result<Option<String>> {
        with_store(|s| -> Result<_> {
            let target_type = s
                .type_as_struct(self.model_type)?
                .data
                .get_prop(&self.field);
            target_type
                .map(get_rel_name)
                .transpose()
                .map(|o| o.flatten())
        })
    }
}

/// Possible cardinalities are:
/// (Optional, Optional): [Left] 0..1 --> 0..1 [Right]
/// (One, Optional): [Left] 1..1 --> 0..1 [Right]
/// (Optional, Many) [Left] 0..1 --> 0..n [Right]
/// (One, Many) [Left] 1..1 --> 0..n [Right]
/// The model on the right will have the foreign key
#[derive(Debug)]
pub struct Relationship {
    pub name: String,
    pub left: RelationshipModel,
    pub right: RelationshipModel,
}

impl TryFrom<(RelationshipModel, RelationshipModel)> for Relationship {
    type Error = TgError;

    fn try_from(pair: (RelationshipModel, RelationshipModel)) -> Result<Self, Self::Error> {
        let (first, second) = pair;
        let (left, right) = match (first.cardinality, second.cardinality) {
            (Cardinality::One, Cardinality::One)
            | (Cardinality::Optional, Cardinality::Optional) => {
                todo!()
            }
            (Cardinality::One, Cardinality::Optional)
            | (Cardinality::One, Cardinality::Many)
            | (Cardinality::Optional, Cardinality::Many) => (first, second),
            (Cardinality::Optional, Cardinality::One)
            | (Cardinality::Many, Cardinality::One)
            | (Cardinality::Many, Cardinality::Optional) => (second, first),
            (Cardinality::Many, Cardinality::Many) => {
                return Err("Many to many relationship not supported".to_string());
            }
        };
        let name = match (left.get_rel_name()?, right.get_rel_name()?) {
            (None, None) => {
                let left_name = with_store(|s| -> Result<_> {
                    s.get_type_name(left.model_type)?
                        .map(|s| s.to_string())
                        .ok_or_else(|| "Prisma model must have explicit name".to_string())
                })?;
                let right_name = with_store(|s| -> Result<_> {
                    s.get_type_name(right.model_type)?
                        .map(|s| s.to_string())
                        .ok_or_else(|| "Prisma model must have explicit name".to_string())
                })?;
                format!("{}To{}", left_name, right_name)
            }
            (Some(name), None) => name,
            (None, Some(name)) => name,
            (Some(name1), Some(name2)) => {
                if name1 == name2 {
                    name1
                } else {
                    return Err("Relationship names do not match".to_string());
                }
            }
        };

        Ok(Relationship { name, left, right })
    }
}

use registry::RelationshipRegistry;

#[cfg(test)]
mod test {
    use crate::errors::Result;
    use crate::global_store::with_store;
    use crate::runtimes::prisma::relationship::registry::RelationshipRegistry;
    use crate::test_utils::*;

    #[test]
    fn test_relationship_discovery() -> Result<(), String> {
        let user = Lib::structb(
            TypeStruct::default()
                .prop(
                    "id",
                    Lib::integerb(TypeInteger::default(), TypeBase::default().as_id())?,
                )
                .prop(
                    "name",
                    Lib::stringb(TypeString::default(), TypeBase::default())?,
                )
                .prop(
                    "posts",
                    Lib::arrayb(
                        TypeArray::of(Lib::proxyb(TypeProxy::new("Post"))?),
                        TypeBase::default(),
                    )?,
                ),
            TypeBase::named("User"),
        )?;

        let post = Lib::structb(
            TypeStruct::default()
                .prop(
                    "id",
                    Lib::integerb(TypeInteger::default(), TypeBase::default().as_id())?,
                )
                .prop(
                    "title",
                    Lib::stringb(TypeString::default(), TypeBase::default())?,
                )
                .prop("author", Lib::proxyb(TypeProxy::new("User"))?),
            TypeBase::named("Post"),
        )?;

        let registry = with_store(|s| -> Result<_> {
            let models = [s.type_as_struct(user)?, s.type_as_struct(post)?];
            let reg = RelationshipRegistry::from(&models)?;
            Ok(reg)
        });

        insta::assert_debug_snapshot!(registry);

        Ok(())
    }
}
