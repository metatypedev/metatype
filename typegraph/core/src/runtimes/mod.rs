// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

pub mod aws;
pub mod deno;
pub mod graphql;
pub mod prisma;
pub mod python;
pub mod random;
pub mod temporal;
pub mod typegate;
pub mod typegraph;
pub mod wasi;

use std::rc::Rc;

use crate::conversion::runtimes::MaterializerConverter;
use crate::global_store::Store;
use crate::runtimes::prisma::migration::{
    prisma_apply, prisma_create, prisma_deploy, prisma_diff, prisma_reset,
};
use crate::runtimes::prisma::with_prisma_runtime;
use crate::runtimes::typegraph::TypegraphOperation;
use crate::validation::types::validate_value;
use crate::wit::aws::S3RuntimeData;
use crate::wit::core::{FuncParams, MaterializerId, RuntimeId, TypeId as CoreTypeId};
use crate::wit::runtimes::{
    self as wit, BaseMaterializer, Error as TgError, GraphqlRuntimeData, HttpRuntimeData,
    MaterializerHttpRequest, PrismaLinkData, PrismaMigrationOperation, PrismaRuntimeData,
    RandomRuntimeData, TemporalOperationData, TemporalRuntimeData,
};
use crate::{typegraph::TypegraphContext, wit::runtimes::Effect as WitEffect};
use enum_dispatch::enum_dispatch;

use self::aws::S3Materializer;
pub use self::deno::{DenoMaterializer, MaterializerDenoImport, MaterializerDenoModule};
pub use self::graphql::GraphqlMaterializer;
use self::prisma::relationship::prisma_link;
use self::prisma::type_generation::replace_variables_to_indices;
use self::prisma::{PrismaMaterializer, PrismaRuntimeContext};
pub use self::python::PythonMaterializer;
pub use self::random::RandomMaterializer;
use self::temporal::temporal_operation;
pub use self::temporal::TemporalMaterializer;
use self::typegate::TypegateOperation;
pub use self::wasi::WasiMaterializer;

type Result<T, E = TgError> = std::result::Result<T, E>;

#[derive(Debug, Clone)]
pub enum Runtime {
    Deno,
    Graphql(Rc<GraphqlRuntimeData>),
    Http(Rc<HttpRuntimeData>),
    Python,
    Random(Rc<RandomRuntimeData>),
    WasmEdge,
    Prisma(Rc<PrismaRuntimeData>, Rc<PrismaRuntimeContext>),
    PrismaMigration,
    Temporal(Rc<TemporalRuntimeData>),
    Typegate,
    Typegraph,
    S3(Rc<S3RuntimeData>),
}

#[derive(Debug, Clone)]
pub struct Materializer {
    pub runtime_id: RuntimeId,
    pub effect: wit::Effect,
    pub data: MaterializerData,
}

impl Materializer {
    fn deno(data: DenoMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id: Store::get_deno_runtime(),
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn graphql(runtime_id: RuntimeId, data: GraphqlMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn http(runtime_id: RuntimeId, data: MaterializerHttpRequest, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn python(runtime_id: RuntimeId, data: PythonMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn random(runtime_id: RuntimeId, data: RandomMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn wasi(runtime_id: RuntimeId, data: WasiMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn prisma(runtime_id: RuntimeId, data: PrismaMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn prisma_migrate(
        runtime_id: RuntimeId,
        data: PrismaMigrationOperation,
        effect: wit::Effect,
    ) -> Self {
        Self {
            runtime_id,
            effect,
            data: data.into(),
        }
    }

    fn temporal(runtime_id: RuntimeId, data: TemporalMaterializer, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: Rc::new(data).into(),
        }
    }

    fn typegate(runtime_id: RuntimeId, data: TypegateOperation, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: data.into(),
        }
    }

    fn typegraph(runtime_id: RuntimeId, data: TypegraphOperation, effect: wit::Effect) -> Self {
        Self {
            runtime_id,
            effect,
            data: data.into(),
        }
    }
}

#[derive(Debug, Clone)]
#[enum_dispatch]
pub enum MaterializerData {
    Deno(Rc<DenoMaterializer>),
    GraphQL(Rc<GraphqlMaterializer>),
    Http(Rc<MaterializerHttpRequest>),
    Python(Rc<PythonMaterializer>),
    Random(Rc<RandomMaterializer>),
    WasmEdge(Rc<WasiMaterializer>),
    Prisma(Rc<PrismaMaterializer>),
    PrismaMigration(PrismaMigrationOperation),
    Temporal(Rc<TemporalMaterializer>),
    Typegate(TypegateOperation),
    Typegraph(TypegraphOperation),
    S3(Rc<S3Materializer>),
}

macro_rules! prisma_op {
    ( $rt:expr, $model:expr, $fn:ident, $name:expr, $effect:expr ) => {{
        let types = with_prisma_runtime($rt, |ctx| ctx.$fn($model.into()))?;

        let mat = PrismaMaterializer {
            table: $crate::types::TypeId($model)
                .type_name()?
                .ok_or_else(|| "prisma model must be named".to_string())?,
            operation: $name.to_string(),
            ordered_keys: None,
        };

        let mat_id = Store::register_materializer(Materializer::prisma($rt, mat, $effect));

        Ok(FuncParams {
            inp: types.input.into(),
            out: types.output.into(),
            mat: mat_id,
        })
    }};

    ( $rt:expr, $model:expr, $fn:ident, $name:expr ) => {
        prisma_op!($rt, $model, $fn, $name, WitEffect::None)
    };
}

impl wit::Runtimes for crate::Lib {
    fn get_deno_runtime() -> RuntimeId {
        Store::get_deno_runtime()
    }

    fn register_deno_func(
        data: wit::MaterializerDenoFunc,
        effect: wit::Effect,
    ) -> Result<wit::MaterializerId> {
        // TODO: check code is valid function?
        let mat = Materializer::deno(DenoMaterializer::Inline(data), effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_deno_static(
        data: wit::MaterializerDenoStatic,
        type_id: CoreTypeId,
    ) -> Result<wit::MaterializerId> {
        validate_value(
            serde_json::from_str::<serde_json::Value>(&data.value).map_err(|e| e.to_string())?,
            type_id.into(),
            "<V>".to_string(),
        )?;

        Ok(Store::register_materializer(Materializer::deno(
            DenoMaterializer::Static(deno::MaterializerDenoStatic {
                value: serde_json::from_str(&data.value).map_err(|e| e.to_string())?,
            }),
            wit::Effect::None,
        )))
    }

    fn get_predefined_deno_func(
        data: wit::MaterializerDenoPredefined,
    ) -> Result<wit::MaterializerId> {
        Store::get_predefined_deno_function(data.name)
    }

    fn import_deno_function(
        data: wit::MaterializerDenoImport,
        effect: wit::Effect,
    ) -> Result<wit::MaterializerId> {
        let module = Store::get_deno_module(data.module);
        let data = MaterializerDenoImport {
            func_name: data.func_name,
            module,
            secrets: data.secrets,
        };
        let mat = Materializer::deno(DenoMaterializer::Import(data), effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_graphql_runtime(data: GraphqlRuntimeData) -> Result<RuntimeId> {
        let runtime = Runtime::Graphql(data.into());
        Ok(Store::register_runtime(runtime))
    }

    fn graphql_query(
        base: BaseMaterializer,
        data: wit::MaterializerGraphqlQuery,
    ) -> Result<wit::MaterializerId> {
        let data = GraphqlMaterializer::Query(data);
        let mat = Materializer::graphql(base.runtime, data, base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn graphql_mutation(
        base: BaseMaterializer,
        data: wit::MaterializerGraphqlQuery,
    ) -> Result<wit::MaterializerId> {
        let data = GraphqlMaterializer::Mutation(data);
        let mat = Materializer::graphql(base.runtime, data, base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_http_runtime(data: wit::HttpRuntimeData) -> Result<wit::RuntimeId, wit::Error> {
        Ok(Store::register_runtime(Runtime::Http(data.into())))
    }

    fn http_request(
        base: wit::BaseMaterializer,
        data: wit::MaterializerHttpRequest,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::http(base.runtime, data, base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_python_runtime() -> Result<wit::RuntimeId, wit::Error> {
        Ok(Store::register_runtime(Runtime::Python))
    }

    fn from_python_lambda(
        base: wit::BaseMaterializer,
        data: wit::MaterializerPythonLambda,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::python(base.runtime, PythonMaterializer::Lambda(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn from_python_def(
        base: wit::BaseMaterializer,
        data: wit::MaterializerPythonDef,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::python(base.runtime, PythonMaterializer::Def(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn from_python_module(
        base: wit::BaseMaterializer,
        data: wit::MaterializerPythonModule,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::python(base.runtime, PythonMaterializer::Module(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn from_python_import(
        base: wit::BaseMaterializer,
        data: wit::MaterializerPythonImport,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::python(base.runtime, PythonMaterializer::Import(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_random_runtime(
        data: wit::RandomRuntimeData,
    ) -> Result<wit::MaterializerId, wit::Error> {
        Ok(Store::register_runtime(Runtime::Random(data.into())))
    }

    fn create_random_mat(
        base: wit::BaseMaterializer,
        data: wit::MaterializerRandom,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat =
            Materializer::random(base.runtime, RandomMaterializer::Runtime(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_wasmedge_runtime() -> Result<wit::RuntimeId, wit::Error> {
        Ok(Store::register_runtime(Runtime::WasmEdge))
    }

    fn from_wasi_module(
        base: wit::BaseMaterializer,
        data: wit::MaterializerWasi,
    ) -> Result<wit::MaterializerId, wit::Error> {
        let mat = Materializer::wasi(base.runtime, WasiMaterializer::Module(data), base.effect);
        Ok(Store::register_materializer(mat))
    }

    fn register_prisma_runtime(data: wit::PrismaRuntimeData) -> Result<wit::RuntimeId, wit::Error> {
        Ok(Store::register_runtime(Runtime::Prisma(
            data.into(),
            Default::default(),
        )))
    }

    fn prisma_find_unique(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, find_unique, "findUnique")
    }

    fn prisma_find_many(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, find_many, "findMany")
    }

    fn prisma_find_first(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, find_first, "findFirst")
    }

    fn prisma_aggregate(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, aggregate, "aggregate")
    }

    fn prisma_group_by(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, group_by, "groupBy")
    }

    fn prisma_count(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(runtime, model, count, "count")
    }

    fn prisma_create_one(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            create_one,
            "createOne",
            WitEffect::Create(false)
        )
    }

    fn prisma_create_many(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            create_many,
            "createMany",
            WitEffect::Create(false)
        )
    }

    fn prisma_update_one(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            update_one,
            "updateOne",
            WitEffect::Update(false)
        )
    }

    fn prisma_update_many(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            update_many,
            "updateMany",
            WitEffect::Update(false)
        )
    }

    fn prisma_upsert_one(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            upsert_one,
            "upsertOne",
            WitEffect::Update(true)
        )
    }

    fn prisma_delete_one(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            delete_one,
            "deleteOne",
            WitEffect::Delete(true)
        )
    }

    fn prisma_delete_many(runtime: RuntimeId, model: CoreTypeId) -> Result<FuncParams, wit::Error> {
        prisma_op!(
            runtime,
            model,
            delete_many,
            "deleteMany",
            WitEffect::Delete(true)
        )
    }

    fn prisma_execute(
        runtime: RuntimeId,
        query: String,
        param: CoreTypeId,
        effect: WitEffect,
    ) -> Result<FuncParams, wit::Error> {
        let types = with_prisma_runtime(runtime, |ctx| ctx.execute_raw(param.into()))?;
        let proc = replace_variables_to_indices(query, types.input)?;
        let mat = PrismaMaterializer {
            table: proc.query,
            operation: "executeRaw".to_string(),
            ordered_keys: Some(proc.ordered_keys),
        };
        let mat_id = Store::register_materializer(Materializer::prisma(runtime, mat, effect));
        Ok(FuncParams {
            inp: types.input.into(),
            out: types.output.into(),
            mat: mat_id,
        })
    }

    fn prisma_query_raw(
        runtime: RuntimeId,
        query: String,
        param: Option<CoreTypeId>,
        out: CoreTypeId,
    ) -> Result<FuncParams, wit::Error> {
        let types = with_prisma_runtime(runtime, |ctx| {
            ctx.query_raw(param.map(|v| v.into()), out.into())
        })?;
        let proc = replace_variables_to_indices(query, types.input)?;
        let mat = PrismaMaterializer {
            table: proc.query,
            operation: "queryRaw".to_string(),
            ordered_keys: Some(proc.ordered_keys),
        };
        let mat_id =
            Store::register_materializer(Materializer::prisma(runtime, mat, WitEffect::None));
        Ok(FuncParams {
            inp: types.input.into(),
            out: types.output.into(),
            mat: mat_id,
        })
    }

    fn prisma_link(data: PrismaLinkData) -> Result<CoreTypeId, wit::Error> {
        let mut builder = prisma_link(data.target_type.into())?;
        if let Some(name) = data.relationship_name {
            builder = builder.name(name);
        }
        if let Some(fkey) = data.foreign_key {
            builder = builder.fkey(fkey);
        }
        if let Some(field) = data.target_field {
            builder = builder.field(field);
        }
        if let Some(unique) = data.unique {
            builder = builder.unique(unique);
        }
        Ok(builder.build()?.into())
    }

    fn prisma_migration(operation: PrismaMigrationOperation) -> Result<FuncParams, wit::Error> {
        use PrismaMigrationOperation as Op;

        let (effect, (inp, out)) = match operation {
            Op::Diff => (WitEffect::None, prisma_diff()?),
            Op::Create => (WitEffect::Create(false), prisma_create()?),
            Op::Apply => (WitEffect::Update(false), prisma_apply()?),
            Op::Deploy => (WitEffect::Update(true), prisma_deploy()?),
            Op::Reset => (WitEffect::Delete(true), prisma_reset()?),
        };

        let mat_id = Store::register_materializer(Materializer::prisma_migrate(
            Store::get_prisma_migration_runtime(),
            operation,
            effect,
        ));

        Ok(FuncParams {
            inp: inp.into(),
            out: out.into(),
            mat: mat_id,
        })
    }

    fn register_temporal_runtime(data: TemporalRuntimeData) -> Result<RuntimeId, wit::Error> {
        Ok(Store::register_runtime(Runtime::Temporal(data.into())))
    }

    fn generate_temporal_operation(
        runtime: RuntimeId,
        data: TemporalOperationData,
    ) -> Result<FuncParams, wit::Error> {
        temporal_operation(runtime, data)
    }

    fn register_typegate_materializer(
        operation: wit::TypegateOperation,
    ) -> Result<MaterializerId, wit::Error> {
        use wit::TypegateOperation as WitOp;
        use TypegateOperation as Op;

        let (effect, op) = match operation {
            WitOp::ListTypegraphs => (WitEffect::None, Op::ListTypegraphs),
            WitOp::FindTypegraph => (WitEffect::None, Op::FindTypegraph),
            WitOp::AddTypegraph => (WitEffect::Create(true), Op::AddTypegraph),
            WitOp::RemoveTypegraph => (WitEffect::Delete(true), Op::RemoveTypegraph),
            WitOp::GetSerializedTypegraph => (WitEffect::None, Op::GetSerializedTypegraph),
        };

        Ok(Store::register_materializer(Materializer::typegate(
            Store::get_typegate_runtime(),
            op,
            effect,
        )))
    }

    fn register_typegraph_materializer(
        operation: wit::TypegraphOperation,
    ) -> Result<MaterializerId, wit::Error> {
        use wit::TypegraphOperation as WitOp;
        use TypegraphOperation as Op;

        let (effect, op) = match operation {
            WitOp::Resolver => (WitEffect::None, Op::Resolver),
            WitOp::GetType => (WitEffect::None, Op::GetType),
            WitOp::GetSchema => (WitEffect::None, Op::GetSchema),
        };

        Ok(Store::register_materializer(Materializer::typegraph(
            Store::get_typegraph_runtime(),
            op,
            effect,
        )))
    }
}
