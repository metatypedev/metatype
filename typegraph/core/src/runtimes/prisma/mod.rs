// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

pub mod errors;
pub mod relationship;
pub mod type_generation;
mod type_utils;

use std::fmt::Debug;

use common::typegraph::runtimes::prisma as cm;
use common::typegraph::Materializer;
use indexmap::IndexMap;

use crate::conversion::runtimes::MaterializerConverter;
use crate::errors::Result;
use crate::global_store::Store;
use crate::typegraph::TypegraphContext;
use crate::wit::runtimes::{self as wit, RuntimeId};

use self::relationship::{Cardinality, Relationship, RelationshipModel};
use self::type_generation::TypeGenContext;

use super::Runtime;

pub struct PrismaRuntimeContext(std::cell::RefCell<Option<TypeGenContext>>);

impl Debug for PrismaRuntimeContext {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let inner = self.0.borrow();
        f.debug_tuple("PrismaRuntimeContext").field(&inner).finish()
    }
}

impl Default for PrismaRuntimeContext {
    fn default() -> Self {
        Self(std::cell::RefCell::new(Some(TypeGenContext::default())))
    }
}

pub fn with_prisma_runtime<R>(
    runtime_id: RuntimeId,
    f: impl FnOnce(&mut TypeGenContext) -> Result<R>,
) -> Result<R> {
    let mut ctx = match Store::get_runtime(runtime_id)? {
        Runtime::Prisma(_, ctx) => Ok(ctx
            .0
            .borrow_mut()
            .take()
            .ok_or_else(|| "prisma runtime context already borrowed".to_string())?),
        _ => Err("not a prisma runtime".to_string()),
    }?;

    let res = f(&mut ctx)?;

    match Store::get_runtime(runtime_id).unwrap() {
        Runtime::Prisma(_, c) => c.0.borrow_mut().replace(ctx),
        _ => unreachable!(),
    };

    Ok(res)
}

#[derive(Debug)]
pub struct PrismaMaterializer {
    pub table: String,
    pub operation: String,
    pub ordered_keys: Option<Vec<String>>,
}

impl MaterializerConverter for PrismaMaterializer {
    fn convert(
        &self,
        c: &mut TypegraphContext,
        runtime_id: RuntimeId,
        effect: wit::Effect,
    ) -> Result<Materializer> {
        let runtime = c.register_runtime(runtime_id)?;
        let mut data = IndexMap::new();
        data.insert(
            "table".to_string(),
            serde_json::Value::String(self.table.clone()),
        );
        data.insert(
            "operation".to_string(),
            serde_json::Value::String(self.operation.clone()),
        );
        if let Some(ordered_keys) = self.ordered_keys.clone() {
            let value = serde_json::to_value(ordered_keys).map_err(|e| e.to_string())?;
            data.insert("ordered_keys".to_string(), value);
        }
        Ok(Materializer {
            name: "prisma_operation".to_string(),
            runtime,
            effect: effect.into(),
            data,
        })
    }
}

pub struct ConversionContext<'a> {
    pub runtime_id: u32,
    pub tg_context: &'a mut TypegraphContext,
}

impl<'a> ConversionContext<'a> {
    pub fn convert_relationship(&mut self, rel: &Relationship) -> Result<cm::Relationship> {
        let left = self.convert_relationship_model(&rel.left)?;
        let right = self.convert_relationship_model(&rel.right)?;
        Ok(cm::Relationship {
            name: rel.name.clone(),
            left,
            right,
        })
    }

    pub fn convert_relationship_model(
        &mut self,
        model: &RelationshipModel,
    ) -> Result<cm::RelationshipModel> {
        Ok(cm::RelationshipModel {
            type_idx: self
                .tg_context
                .register_type(model.model_type, Some(self.runtime_id))?
                .into(),
            field: model.field.clone(),
            cardinality: model.cardinality.into(),
        })
    }
}

impl From<Cardinality> for cm::Cardinality {
    fn from(cardinality: Cardinality) -> Self {
        match cardinality {
            Cardinality::Optional => cm::Cardinality::Optional,
            Cardinality::One => cm::Cardinality::One,
            Cardinality::Many => cm::Cardinality::Many,
        }
    }
}
