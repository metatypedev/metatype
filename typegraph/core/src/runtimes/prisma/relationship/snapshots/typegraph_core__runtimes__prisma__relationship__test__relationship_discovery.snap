---
source: typegraph/core/src/runtimes/prisma/relationship/mod.rs
expression: registry
---
Ok(
    RelationshipRegistry {
        models: {
            4: {
                "posts": "UserToPost",
            },
            7: {
                "author": "UserToPost",
            },
        },
        relationships: {
            "UserToPost": Relationship {
                name: "UserToPost",
                left: RelationshipModel {
                    model_type: 4,
                    wrapper_type: 4,
                    cardinality: One,
                    field: "posts",
                },
                right: RelationshipModel {
                    model_type: 7,
                    wrapper_type: 3,
                    cardinality: Many,
                    field: "author",
                },
            },
        },
    },
)
